### ISSUE TITLE
Incorrect character width calculation

### ISSUE THREAD

**Issue by char101** [jupyter/qtconsole#437]
Hi,

In [console_widget.py#L491](https://github.com/jupyter/qtconsole/blob/master/qtconsole/console_widget.py#L491) the character width is calculated using `boundingRect`. In my system (Windows 10, python 3.8.5, pyqt5  5.15.0) the result of `boundingRect` `width()` is not consistent (even though the font is monospaced).

```python
print(' ', font_metrics.boundingRect(' ').width(), font_metrics.horizontalAdvance(' '), file=f)
for c in range(ord('a'), ord('z') + 1):
  print(chr(c), font_metrics.boundingRect(chr(c)).width(), font_metrics.horizontalAdvance(chr(c)), file=f)
```

```
  1 6 <- ' ' character
a 5 6
b 6 6
c 5 6
d 5 6
e 5 6
f 5 6
g 5 6
h 5 6
i 5 6
j 4 6
k 6 6
l 5 6
m 6 6
n 5 6
o 6 6
p 6 6
q 5 6
r 5 6
s 5 6
t 5 6
u 5 6
v 6 6
w 6 6
x 5 6
y 6 6
z 5 6
```

This results in the wrong window width. What works is to use `horizontalAdvance` which is the alternative to the deprecated `width` method (https://doc.qt.io/qt-5/qfontmetricsf-obsolete.html#width-1).

**Comment by gadiyar** [comment]
+1

I could replicate this issue on upgrading to QtConsole 4.7.6 on Windows 10 - the QtConsole window size was smaller.

Changing to horizontalAdvance(' ') fixed it for me.

**Comment by ccordoba12** [comment]
@char101, could you submit a PR with your fix? Thanks!

**Comment by char101** [comment]
Created https://github.com/jupyter/qtconsole/pull/438

**Comment by yannick1974** [comment]
Hi,

I think it's worth replacing the other `boundingRect`. On my system, for a reason I don't understand, `boundingRect(' ').width()` returns 0 (I tried with different fonts). This leads to a `ZeroDivisionError`  at [`completion_html.py:313`](https://github.com/jupyter/qtconsole/blob/master/qtconsole/completion_html.py#L313) that makes qtconsole crash when using pyqt5 (but not with pyside). Also, another effect is the qtconsole does not honour `c.JupyterWidget.console_width `.

I replaced all the `boundingRect(' ').width()`  by `horizontalAdvance(' ')` and now everything works as expected.
